### PR TITLE
Fix update of CustomButton to save ae_attributes

### PR DESCRIPTION
**Pull request by ZitaNemeckova**
Problem: API do not save custom attributes from UI.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1793556

@miq-bot add_label wip, bug, ivanchuk/no

### CODE REVIEW

**Comment by himdel** [comment]
(spec failure unrelated - https://github.com/ManageIQ/manageiq-api/pull/729)

**Comment by ZitaNemeckova** [comment]
@miq-bot remove_label wip

@lpichler please have a look, thanks :)

File: app/controllers/api/custom_buttons_controller.rb [hunk 1]
```diff
@@ -16,8 +16,13 @@ def create_resource(_type, _id, data)
     def edit_resource(type, id, data)
       custom_button = fetch_custom_button(type, id)
       updated_data = data['resource'] || data
-      updated_data['resource_action'] = find_or_create_resource_action(updated_data['resource_action']) if updated_data.key?('resource_action')
+      resource_action = updated_data["resource_action"]
+      updated_data.except!("resource_action")
       custom_button.update!(updated_data.deep_symbolize_keys) if data.present?
+      if resource_action.present?
+        custom_button.resource_action = find_or_create_resource_action(resource_action)
+        custom_button.save
```
Comment (d-m-u) [inline app/controllers/api/custom_buttons_controller.rb]: Hey Zita, sorry for the drive-by but any chance this could be a `save!` just in case it fails?

Comment (ZitaNemeckova) [reply app/controllers/api/custom_buttons_controller.rb]: @d-m-u Nice catch 👍 Thanks. Fixed :)

File: spec/requests/custom_buttons_spec.rb [hunk 2]
```diff
@@ -131,7 +131,7 @@
       request = {
         'action'    => 'edit',
         'resources' => [
-          { 'id' => cb.id.to_s, 'resource' => {'name' => 'updated 1', 'resource_action' => {'ae_namespace' => 'SYSTEM2'}, 'visibility' => {'roles' => ['_ALL_']}}},
+          { 'id' => cb.id.to_s, 'resource' => {'name' => 'updated 1', 'resource_action' => {'ae_namespace' => 'SYSTEM2', :ae_attributes => {"attribute" => "is present"}}, 'visibility' => {'roles' => ['_ALL_']}}},
```
Comment (lpichler) [inline spec/requests/custom_buttons_spec.rb]: Maybe `:ae_attributes` => `'ae_attributes'` ? - as we have string keys here and we have also keys strings in api request.

File: spec/manageiq-api [hunk 3]
```diff
@@ -0,0 +1 @@
+/Users/zita/Desktop/ManageIQ/manageiq-api/
```
Comment (lpichler) [inline spec/manageiq-api]: ✂️ ?

File: app/controllers/api/custom_buttons_controller.rb [hunk 4]
```diff
@@ -14,10 +14,15 @@ def create_resource(_type, _id, data)
...
     def edit_resource(type, id, data)
+      return if data.empty?
       custom_button = fetch_custom_button(type, id)
       updated_data = data['resource'] || data
-      updated_data['resource_action'] = find_or_create_resource_action(updated_data['resource_action']) if updated_data.key?('resource_action')
-      custom_button.update!(updated_data.deep_symbolize_keys) if data.present?
+      resource_action = updated_data['resource_action']
+      create_or_update_resource_action(updated_data['resource_action']) if resource_action.present?
+      updated_data.except!('resource_action')
+      custom_button.update!(updated_data.deep_symbolize_keys) if updated_data.present?
```
Comment (ZitaNemeckova) [inline app/controllers/api/custom_buttons_controller.rb]: @lpichler This line will change `ae_attributes` in `custom_button.resource_action` from :
```ruby
#<ResourceAction:0x00007fc0dff5fc10
 id: 7603,
 action: nil,
 dialog_id: nil,
 resource_id: 70,
 resource_type: "CustomButton",
 created_at: Wed, 19 Feb 2020 08:45:52 UTC +00:00,
 updated_at: Wed, 19 Feb 2020 13:36:23 UTC +00:00,
 ae_namespace: "SYSTEM",
 ae_class: "PROCESS",
 ae_instance: "Request",
 ae_message: "create",
 ae_attributes: {:pepek=>"VALUE", :KAREL=>"VALUES", :request=>"PEPEK", :service_template=>nil},
 configuration_template_id: nil,
 configuration_template_type: nil>
```
to this:
```ruby
#<ResourceAction:0x00007fc0dff5fc10
 id: 7603,
 action: nil,
 dialog_id: nil,
 resource_id: 70,
 resource_type: "CustomButton",
 created_at: Wed, 19 Feb 2020 08:45:52 UTC +00:00,
 updated_at: Wed, 19 Feb 2020 13:36:52 UTC +00:00,
 ae_namespace: "SYSTEM",
 ae_class: "PROCESS",
 ae_instance: "Request",
 ae_message: "create",
 ae_attributes: {:request=>"PEPEK", :service_template=>nil, :hosts=>nil},
 configuration_template_id: nil,
 configuration_template_type: nil>
```
`custom_button.save` works as expected. Any ideas what's wrong?

File: spec/manageiq-api (hunk 3 above)
Comment (ZitaNemeckova) [reply spec/manageiq-api]: Yes ✂️

File: app/controllers/api/custom_buttons_controller.rb [hunk 5]
```diff
@@ -4,20 +4,24 @@ def create_resource(_type, _id, data)
...
-      custom_button.visibility = data["visibility"].deep_symbolize_keys if data["visibility"]
-      custom_button.resource_action = find_or_create_resource_action(data["resource_action"]) if data.key?("resource_action")
-      if custom_button.save
-        custom_button
-      else
-        raise BadRequestError, "Failed to create new custom button - #{custom_button.errors.full_messages.join(", ")}"
-      end
+      custom_button.save!
+      custom_button.resource_action.update!(data["resource_action"].deep_symbolize_keys) if data.key?("resource_action")
+      custom_button.save!
```
Comment (himdel) [inline app/controllers/api/custom_buttons_controller.rb]: Maybe these should be wrapped in a transaction, so that we can rollback on exception?

Comment (ZitaNemeckova) [reply app/controllers/api/custom_buttons_controller.rb]: Fixed :)

**Comment by ZitaNemeckova** [comment]
@miq-bot add_label ivanchuk/yes
@miq-bot remove_label ivanchuk/no

File: app/controllers/api/custom_buttons_controller.rb [hunk 6]
```diff
@@ -1,24 +1,32 @@
...
-      custom_button.userid = User.current_user.userid
-      custom_button.options = data["options"].deep_symbolize_keys if data["options"]
-      custom_button.visibility = data["visibility"].deep_symbolize_keys if data["visibility"]
-      custom_button.resource_action = find_or_create_resource_action(data["resource_action"]) if data.key?("resource_action")
-      if custom_button.save
+      CustomButton.transaction do
+        custom_button = CustomButton.new(data.except("resource_action", "options"))
+        custom_button.userid = User.current_user.userid
+        custom_button.options = data["options"].deep_symbolize_keys if data["options"]
+        custom_button.save!
```
Comment (lpichler) [inline app/controllers/api/custom_buttons_controller.rb]: do we need to call `save!` here also ?

**Comment by lpichler** [comment]
@ZitaNemeckova Looks good and I had one question about `save!`

File: app/controllers/api/custom_buttons_controller.rb (hunk 6 above)
Comment (ZitaNemeckova) [reply app/controllers/api/custom_buttons_controller.rb]: Seems not. Removed :)

**Comment by ZitaNemeckova** [comment]
I removed `save!` bc it wasn't needed.

**Comment by simaishi** [comment]
@ZitaNemeckova I only need to backport this PR for the referenced BZ, not the other 2 PRs marked as 'related' ?

**Comment by ZitaNemeckova** [comment]
@simaishi It needs those two(https://github.com/ManageIQ/manageiq-ui-classic/pull/6690 and
https://github.com/ManageIQ/manageiq-ui-classic/pull/6695) as well. I added them to the BZ. Thanks :)

**Comment by simaishi** [comment]
Ivanchuk backport details:

```
$ git log -1
commit c527e8373c11ab174a2a15df71295f7ea511f71c
Author: Libor Pichler <lpichler@redhat.com>
Date:   Mon Feb 24 16:04:12 2020 +0100

    Merge pull request #728 from ZitaNemeckova/fix_custom_button_update

    Fix update of CustomButton to save ae_attributes

    (cherry picked from commit ea2365cb085aa8618da596e28402e9ec6e39ce6b)

    https://bugzilla.redhat.com/show_bug.cgi?id=1793556
```

**Pull request merged**